pkg/ingester/errors.go

```diff
@@ -78,111 +70,230 @@ func (e errorWithStatus) GRPCStatus() *status.Status {
 }
 
 // wrapOrAnnotateWithUser prepends the given userID to the given error.
-// If the error is safe, the returned error retains a reference to the former.
+// If the given error matches one of the errors from this package, the
+// returned error retains a reference to the former.
 func wrapOrAnnotateWithUser(err error, userID string) error {
-	// If this is a safe error, we wrap it with userID and return it, because
-	// it might contain extra information for gRPC and our logging middleware.
-	var safe safeToWrap
-	if errors.As(err, &safe) {
+	// If err matches one of the errors from this package,
+	// we wrap it with userID and return it.
+	if errors.As(err, &unavailableError{}) ||
+		errors.As(err, &tsdbUnavailableError{}) ||
+		errors.As(err, &instanceLimitReachedError{}) ||
+		errors.As(err, &validationError{}) ||
+		errors.As(err, &perUserLimitReachedError{}) ||
+		errors.As(err, &perMetricLimitReachedError{}) {
```

Comment:
Each one of the calls like `errors.As(err, &tsdbUnavailableError{})` creates an error that escapes to heap. So if the error is `perMetricLimitReachedError{}` we need to allocate 6 other errors on the heap.

I would rever this change and add the `safeToWrap()` method to all those types instead.